api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2383,6 +2386,225 @@ public void copyObject(String bucketName, String objectName, Map<String,String>
+    throws InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException, IOException,
+    InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
+    InternalException, InvalidArgumentException, InvalidResponseException {
+    if (bucketName == null) {
+      throw new InvalidArgumentException("Destination bucket name cannot be empty");
+    }
+
+    if (objectName == null) {
+      throw new InvalidArgumentException("Destination object name cannot be empty");
+    }
+
+    if (sources.isEmpty()) {
+      throw new InvalidArgumentException("compose sources cannot be empty");
+    }
+
+    checkWriteRequestSse(sse);
+
+    long objectSize = 0;
+    int partsCount = 0;
+    for (int i = 0; i < sources.size(); i++) {
+      ComposeSource src = sources.get(i);
+
+      checkReadRequestSse(src.sse());
+
+      ObjectStat stat = statObject(src.bucketName(), src.objectName(), src.sse());
+      src.buildHeaders(stat.length(), stat.etag());
+
+      if (i != 0 && src.headers().containsKey("x-amz-meta-x-amz-key")) {
+        throw new InvalidArgumentException("Client side encryption is not supported for more than one source");
+      }
+
+      long size = stat.length();
+      if (src.length() != null) {
+        size = src.length();
+      } else {
+        if (src.offset() != null) {
+          size -= src.offset();
+        }
+      }
+
+      if (size < MIN_MULTIPART_SIZE && sources.size() != 1 && i != (sources.size() - 1)) {
+        throw new InvalidArgumentException("source " + src.bucketName() + "/" + src.objectName() + ": size "
+          + size + " must be greater than " + MIN_MULTIPART_SIZE);
+      }
+
+      objectSize += size;
+      if (objectSize > MAX_OBJECT_SIZE) {
+        throw new InvalidArgumentException("Destination object size must be less than " + MAX_OBJECT_SIZE);
+      }
+
+      if (size > MAX_PART_SIZE) {
+        long count = size / MAX_PART_SIZE;
+        long lastPartSize = size - (count * MAX_PART_SIZE);
+        if (lastPartSize > 0) {
+          count++;
+        } else {
+          lastPartSize = MAX_PART_SIZE;
+        }
+
+        if (lastPartSize < MIN_MULTIPART_SIZE && sources.size() != 1 && i != (sources.size() - 1)) {
+          throw new InvalidArgumentException("source " + src.bucketName() + "/" + src.objectName() + ": "
+            + "for multipart split upload of " + size
+            + ", last part size is less than " + MIN_MULTIPART_SIZE);
+        }
+
+        partsCount += (int)count;
+      } else {
+        partsCount++;
+      }
+
+      if (partsCount > MAX_MULTIPART_COUNT) {
+        throw new InvalidArgumentException("Compose sources create more than allowed multipart count "
+          + MAX_MULTIPART_COUNT);
+      }
+    }
+
+    if (partsCount == 1) {
+      ComposeSource src = sources.get(0);
+      copyObject(bucketName, objectName, headerMap, sse, src.bucketName(), src.objectName(), src.sse(),
```

Comment:
If `src.offset() != null || src.length() != null`, you would need to handle add `x-amz-copy-source-range` appropriately.

Reply:
done.